### ISSUE TITLE
Problem with Method in derived class when using `<K extends keyof this>`

### ISSUE THREAD

**Issue by dardino** [Microsoft/TypeScript#20028]
**TypeScript Version:**  2.6.1

I wrote these 2 classes:
**Code**

```ts
class Greeter {
    protected rawProps: any = {};

    public Key: string;

    public GetRawPropValue<K extends keyof this>(p: K): this[K] {
        return this.rawProps[p] as this[K];
    };
    public RawPropExists(p: keyof this): boolean {
        return this.rawProps[p] !== null && this.rawProps[p] !== undefined;
     };
    public RawPropExistsWrong<K extends keyof this>(p: K): boolean {
        return this.rawProps[p] !== null && this.rawProps[p] !== undefined;
     };

    constructor() {

    }
}

class DerivedGreeter extends Greeter { 
    public Description: string;
}
```

originally I wrote the `RawPropExists` with the syntax of `RawPropExistsWrong`
but I got an unexpected error when calling it in derived class `DerivedGreeter`

**Expected behavior:**
```typescript
// greeter
let gr = new Greeter;
let e4 = gr.RawPropExistsWrong("Key"); // OK
// derived greeter
let dg = new DerivedGreeter;
let e1 = dg.RawPropExists("Description"); // OK
let rv = dg.GetRawPropValue("Description"); // OK
let e2 = dg.RawPropExistsWrong("Description"); // OK, I think there is no reason to give an error
let e3 = dg.RawPropExistsWrong("Key"); // OK
```

**Actual behavior:**
```typescript
// greeter
let gr = new Greeter;
let e4 = gr.RawPropExistsWrong("Key"); // OK
// derived greeter
let dg = new DerivedGreeter;
let e1 = dg.RawPropExists("Description"); // OK
let rv = dg.GetRawPropValue("Description"); // OK
let e2 = dg.RawPropExistsWrong("Description"); // Error: Argument of type '"Description"' is not assignable to parameter of type 'keyof this'. Type '"Description"' is not assignable to type '"GetRawPropValue" | "RawPropExists" | "RawPropExistsWrong" | "Key"'.
let e3 = dg.RawPropExistsWrong("Key"); // OK
```

**Comment by dardino** [comment]
I think is similar to this issue: #19961

**Comment by alexrock** [comment]
Not the same. This issue is fixed in master #19655

**Comment by sandersn** [comment]
This issue is also fixed in master. Duplicate of #16930 -- `this` in type parameters should have caused `this` to be instantiated as DerivedGreeter, but did not.
